### PR TITLE
First implementation of SM4

### PR DESCRIPTION
**Pull request by pablooliveira**
SM4: Sherman Morrison, mix between SM3 + SM2
Leave zero denominators for later (SM3), and when none are left then split (SM2)

### CODE REVIEW

**Comment by pablooliveira** [comment]
@PurplePachyderm and @fmgjcoppens a simple implementation of the SM4 hybrid version we had discussed.

**Comment by pablooliveira** [comment]
The implementation is simple:
- First it tries to find any working order
- Once all the remaining updates are failing it changes the strategy to splitting for the remaining updates

Other variants are possible such as splitting only once and then doing back and forths between SM2 and SM3.

I think there are very few cases when this will happen; so this simple strategy is enough to evaluate the interest for an hybrid version I think.

File: tests/vfc_test_h5.cpp [hunk 1]
```diff
@@ -137,6 +137,8 @@ int test_cycle(H5File file, int cycle, std::string version, vfc_probes * probes)
     SM2(slater_inverse, dim, nupdates, u, col_update_index);
   } else if (version == "sm3") {
     SM3(slater_inverse, dim, nupdates, u, col_update_index);
+  } else if (version == "sm3") {
```
Comment (fmgjcoppens) [inline tests/vfc_test_h5.cpp]: I suppose this is a copy/paste typo and needs to be 'sm4'.

**Review by fmgjcoppens — CHANGES_REQUESTED** [review]
I suppose in 'tests/vfc_test_h5.cpp' on line 140 'sm3' needs to be changed to 'sm4'?

File: tests/vfc_test_h5.cpp (hunk 1 above)
Comment (pablooliveira) [reply tests/vfc_test_h5.cpp]: Just did it a few pico-seconds ago. Thanks for the catch ;-)

**Review by fmgjcoppens — APPROVED** [review]

**Pull request merged**